scipy/linalg/tests/test_lapack.py

```diff
@@ -3298,3 +3298,105 @@ def test_gges_tgsen(dtype):
 
     assert_allclose(s[0, 0] / t[0, 0], d2, rtol=0, atol=atol)
     assert_allclose(s[1, 1] / t[1, 1], d1, rtol=0, atol=atol)
+
+
+@pytest.mark.parametrize(
+    "a, b, c, d, e, f, rans, lans",
+    [(np.array([[4.0,   1.0,  1.0,  2.0],
+                [0.0,   3.0,  4.0,  1.0],
+                [0.0,   1.0,  3.0,  1.0],
+                [0.0,   0.0,  0.0,  6.0]]),
+      np.array([[1.0,   1.0,  1.0,  1.0],
+                [0.0,   3.0,  4.0,  1.0],
+                [0.0,   1.0,  3.0,  1.0],
+                [0.0,   0.0,  0.0,  4.0]]),
+      np.array([[-4.0,  7.0,  1.0, 12.0],
+                [-9.0,  2.0, -2.0, -2.0],
+                [-4.0,  2.0, -2.0,  8.0],
+                [-7.0,  7.0, -6.0, 19.0]]),
+      np.array([[2.0,   1.0,  1.0,  3.0],
+                [0.0,   1.0,  2.0,  1.0],
+                [0.0,   0.0,  1.0,  1.0],
+                [0.0,   0.0,  0.0,  2.0]]),
+      np.array([[1.0,   1.0,  1.0,  2.0],
+                [0.0,   1.0,  4.0,  1.0],
+                [0.0,   0.0,  1.0,  1.0],
+                [0.0,   0.0,  0.0,  1.0]]),
+      np.array([[-7.0,  5.0,  0.0,  7.0],
+                [-5.0,  1.0, -8.0,  0.0],
+                [-1.0,  2.0, -3.0,  5.0],
+                [-3.0,  2.0,  0.0,  5.0]]),
+      np.array([[1.0,   1.0,  1.0,  1.0],
+                [-1.0,  2.0, -1.0, -1.0],
+                [-1.0,  1.0,  3.0,  1.0],
+                [-1.0,  1.0, -1.0,  4.0]]),
+      np.array([[4.0,  -1.0,  1.0, -1.0],
+                [1.0,   3.0, -1.0,  1.0],
+                [-1.0,  1.0,  2.0, -1.0],
+                [1.0,  -1.0,  1.0,  1.0]]))])
+@pytest.mark.parametrize('dtype', REAL_DTYPES)
+def test_tgsyl_NAG(a, b, c, d, e, f, rans, lans, dtype):
+    atol = 1e-4
+
+    tgsyl = get_lapack_funcs(('tgsyl'), dtype=dtype)
+    rout, lout, scale, dif, info = tgsyl(a, b, c, d, e, f)
+
+    assert_equal(info, 0)
+    assert_allclose(scale, 1.0, rtol=0, atol=np.finfo(dtype).eps*100,
+                    err_msg="SCALE must be 1.0")
+    assert_allclose(dif, 0.0, rtol=0, atol=np.finfo(dtype).eps*100,
+                    err_msg="DIF must be nearly 0")
+    assert_allclose(rout, rans, atol=atol,
+                    err_msg="Solution for R is incorrect")
+    assert_allclose(lout, lans, atol=atol,
+                    err_msg="Solution for L is incorrect")
+
+
+@pytest.mark.parametrize('dtype', REAL_DTYPES)
+@pytest.mark.parametrize('trans', ('N', 'T'))
+@pytest.mark.parametrize('ijob', [0, 1, 2, 3, 4])
+def test_tgsyl(dtype, trans, ijob):
+
+    seed(2023)
+    atol = 1e-2
+    if dtype == np.float32:
+        atol = 5e-2
```

Comment:
I'm still not too sure why the tolerance needs to be this low. Maybe there should be some investigation why this is happening?

Reply:
Just to be on the same page, I think you mean the tolerance is too high (too loose). I agree. I think the only thing I can think of checking here would be if we are loosing precision between fortran lapack output and python. The fortran wrapper is being constructed by f2py and I am not sure how to debug it directly. Will write statements in the template signature file work? Or is there some more sophisticated debug option I am not aware of? @ilayn Please also weigh in if you have thoughts.

Reply:
Alternatively, I thought of using the same tests Lapack has. I see they test stgsyl here: https://github.com/Reference-LAPACK/lapack/blob/f86d08bae6c46685e0f759850e82dc0a8fa00a21/TESTING/EIG/serrgg.f#LL1131C4-L1131C4
But it looks like they are using 1x1 matrices or testing error with dimensions maybe.

Reply:
It might be worth doing. But, if it's needed, can probably be done in another PR. TBH I'm not a LAPACK expert so I'm going to defer to what @ilayn thinks.

Reply:
I have one proper idea and one hunch, because we are getting data from random arrays, their spectrum is probably too close to have a decent solution hence could you please add `5*np.eye(m, dtype=dtype)` to `a` and subtract the same from `d`? 

Also (hunch) could you please set `scale` to `1` in the wrapper? I feel like it gets referenced without being set. If locally it passes with proper tolerance values (around `1e-10` should be ok for double precision) then it is OK.

Reply:
@ilayn 
> could you please add 5*np.eye(m, dtype=dtype) to a and subtract the same from d?

Thanks for your insight! I tried this and on my system it improved things a lot. I could go down to any atol for double precision (1e-10 works). For single precision atol =2e-3 works but 1e-3 fails. From what I have seen other systems used  in the CI here need somewhat looser tolerances than my system. I'll try with 1e-10 and 5e-3 for my next checkin and see how it goes.

> Also (hunch) could you please set scale to 1 in the wrapper?

Currently, the intent for the scale argument in the wrapper is out. Do you want me to make the intent in,out or inout? Will wait for your response before making changes